From 2f032f32fa189c5f6cb7d5b2ede94166f2cdac26 Mon Sep 17 00:00:00 2001
From: Axel Huebl <axel.huebl@plasma.ninja>
Date: Fri, 27 Mar 2020 00:55:11 -0700
Subject: [PATCH 1/2] CMake: Require Only Needed MPI Component

From the `FindMPI.cmake` module:
```
The module exposes the components ``C``, ``CXX``, ``MPICXX`` and ``Fortran``.
Each of these controls the various MPI languages to search for.
The difference between ``CXX`` and ``MPICXX`` is that ``CXX`` refers to the
MPI C API being usable from C++, whereas ``MPICXX`` refers to the MPI-2 C++ API
that was removed again in MPI-3.
```
---
 CHANGELOG.rst  | 2 ++
 CMakeLists.txt | 8 +++-----
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/CHANGELOG.rst b/CHANGELOG.rst
index d87e1a527..14d408ed8 100644
--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -23,6 +23,8 @@ Bug Fixes
 Other
 """""
 
+- CMake: Require only C-in-CXX MPI component #710
+
 
 0.13.1
 ------
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 796e0addf..e05d0bda0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -133,14 +133,14 @@ option(openPMD_BUILD_EXAMPLES  "Build the examples" ${BUILD_EXAMPLES})
 #
 # external library: MPI (optional)
 if(openPMD_USE_MPI STREQUAL AUTO)
-    find_package(MPI)
+    find_package(MPI COMPONENTS CXX)
     if(MPI_FOUND)
         set(openPMD_HAVE_MPI TRUE)
     else()
         set(openPMD_HAVE_MPI FALSE)
     endif()
 elseif(openPMD_USE_MPI)
-    find_package(MPI REQUIRED)
+    find_package(MPI REQUIRED COMPONENTS CXX)
     set(openPMD_HAVE_MPI TRUE)
 else()
     set(openPMD_HAVE_MPI FALSE)
@@ -438,9 +438,7 @@ if(openPMD_BUILD_TESTING)
 endif()
 
 if(openPMD_HAVE_MPI)
-    # MPI targets: CMake 3.9+
-    # note: often the PUBLIC dependency to CXX is missing in C targets...
-    target_link_libraries(openPMD PUBLIC MPI::MPI_C MPI::MPI_CXX)
+    target_link_libraries(openPMD PUBLIC MPI::MPI_CXX)
 endif()
 
 # JSON Backend and User-Facing Runtime Options

From 0251fe96c6d38085866720e1dc6ecea39b7419da Mon Sep 17 00:00:00 2001
From: Axel Huebl <axel.huebl@plasma.ninja>
Date: Fri, 27 Mar 2020 14:06:51 -0700
Subject: [PATCH 2/2] Clang+MPI: MPI::MPI_C Target Required

... and `enable_language(C)` ...
---
 CMakeLists.txt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index e05d0bda0..6c55a51f3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -133,14 +133,14 @@ option(openPMD_BUILD_EXAMPLES  "Build the examples" ${BUILD_EXAMPLES})
 #
 # external library: MPI (optional)
 if(openPMD_USE_MPI STREQUAL AUTO)
-    find_package(MPI COMPONENTS CXX)
+    find_package(MPI COMPONENTS C CXX)
     if(MPI_FOUND)
         set(openPMD_HAVE_MPI TRUE)
     else()
         set(openPMD_HAVE_MPI FALSE)
     endif()
 elseif(openPMD_USE_MPI)
-    find_package(MPI REQUIRED COMPONENTS CXX)
+    find_package(MPI REQUIRED COMPONENTS C CXX)
     set(openPMD_HAVE_MPI TRUE)
 else()
     set(openPMD_HAVE_MPI FALSE)
@@ -438,7 +438,7 @@ if(openPMD_BUILD_TESTING)
 endif()
 
 if(openPMD_HAVE_MPI)
-    target_link_libraries(openPMD PUBLIC MPI::MPI_CXX)
+    target_link_libraries(openPMD PUBLIC MPI::MPI_C MPI::MPI_CXX)
 endif()
 
 # JSON Backend and User-Facing Runtime Options
